### DISCUSSION TITLE
Does `react-three-rapier` support on-demand rendering?

### DISCUSSION THREAD

**Discussion by jlarmstrongiv** [pmndrs/react-three-rapier#305] [Q&A] [answered]
Here’s the [docs](https://docs.pmnd.rs/react-three-fiber/advanced/scaling-performance#on-demand-rendering) in `react-three-fiber`

Here’s [reference](https://github.com/pmndrs/react-three-fiber/discussions/1884) to support in `react-spring`

Here’s a [reference](https://github.com/pmndrs/use-cannon/blob/master/packages/react-three-cannon-examples/src/demos/demo-Trimesh.tsx#L103) to support with `react-three-cannon`

I couldn’t find references in `react-three-rapier`, so I’m curious if it’s supported or not.  Thanks for creating such a cool library!

**Comment by wiledal** [comment, marked as answer]
Hey @jlarmstrongiv,

Out of the box, the "demand" frameloop strategy will cause some unexpected behavior, mainly due to the physics stepping function using the delta between frames to calculate physics steps, which will cause physics bodies to move at a faster speed than you might want.

So to achieve a complete manual step, you'd need to pause the physics world and utilize the `step` function provided in `useRapier()`.

Here's a small example, using `invalidate` and `step()`
https://codesandbox.io/s/manual-stepping-demand-frameloop-m429m4

**Reply by jlarmstrongiv** [reply]
Thank you for clarifying @wiledal !  I thought that [manual stepping](https://github.com/pmndrs/react-three-rapier#manual-stepping) might be part of the solution—the example is very helpful!  

In your example, you have the TorusKnot falling onto the plane.  Is there a way to automatically call invalidate only while Collider positions are changing or impulses are applied?

**Reply by wiledal** [reply]
Interesting idea! If you were to run your own stepping loop for the `rapier` simulation, you could trigger `invalidate` after catching the collision with `onCollisionEnter`, thus only rendering the frame where the collision happened.

```tsx
<RigidBody onCollisionEnter={
  // trigger something, like invalidate
}>
  ...
</RigidBody>
```

**Reply by wiledal** [reply]
Oh sorry I read that too fast. I get what you're getting at now. Yes, you could do that by checking if there are any "awake" objects in the scene and only triggering the `invalidate` then.

```
<RigidBody onWake={...} onSleep={...} />
```

For something that explicit though, I'd probably access the raw world api and run sleeping checks on all the RigidBodies in the scene instead.

**Reply by jlarmstrongiv** [reply]
That’s interesting, would I be able to do this on a larger scale with everything in the physics world?  I was googling a bit and saw that there is an `IslandManager` that can detect if a rigid body moved recently.  

The use-case I’m thinking of is having an animation on the top of the page that plays (continuously calls `invalidate`) until everything stops moving.  Either the user’s mouse/movements have stopped (no enemies or other interactivity on screen), or say all the TorusKnots have fallen to their final resting place on the plane.  

What are your thoughts?  Do you have more pointers on where to look next or the best way to move forward?

**Reply by jlarmstrongiv** [reply]
Ahh, GitHub replies aren’t syncing right for me.  Followed up on https://discord.com/channels/740090768164651008/969941082144124959/1072463964191080529
